csharp/src/Google.Protobuf/FieldMaskTree.cs

```diff
@@ -356,12 +356,12 @@ public void Merge(IMessage source, IMessage destination, FieldMask.MergeOptions
                     }
                     else
                     {
-                        if (sourceField != null
-                            || !options.ReplacePrimitiveFields)
+                        if ((!field.HasPresence && sourceField != null)
```

Comment:
I'm confused about how sourceField could ever be null here to start with:

- The field is a singular primitive field
- All primitive fields are either value types, or `string`/`ByteString` with validation to ensure they're never null

I could see an *extension* field having a null value, but we don't use in field masks IIRC.

Do you understand what this code was trying to do before? If we can assume it's not null, we can probably write this code (and a comment!) to be more easily understandable. (If we want to protect against nulls even if they're really unexpected, we could have that in another condition - either to throw a useful exception or to perform any suitable action.)

Reply:
I do not understand why this code was checking for null here at all - I was just trying to maintain previous behavior in case there was a case I wasn't seeing.

My initial assumption was that optional fields in protobuf would generate C# `Nullable<_>` properties in the CLR, which seems to be the original assumption made by the implementation, but even this is completely out of conformance with the [field_presence.md doc under Considerations for Merging](https://github.com/protocolbuffers/protobuf/blob/main/docs/field_presence.md#considerations-for-merging):

> Under the no presence rules, it is effectively impossible for a target field to merge-from its default value (using the protobuf's API merging functions). This is because default values are skipped, similar to the no presence serialization discipline. Merging only updates the target (merged-to) message using the non-skipped values from the update (merged-from) message.
> The difference in merging behavior has further implications for protocols which rely on partial "patch" updates. If field presence is not tracked, then an update patch alone cannot represent an update to the default value, because only non-default values are merged-from.
> Updating to set a default value in this case requires some external mechanism, such as FieldMask. However, if presence is tracked, then all explicitly-set values -- even default values -- will be merged into the target.

The only way to get this to work according to spec is to check for (protobuf!) default values if the field is a "no presence" field, and do presence checking otherwise.

And frankly, message fields should do presence checking the same way instead of null-checking (they all are presence fields).

I'll push an update.